data/tutorials/language/1ms_01_functors.md

```diff
@@ -2,189 +2,522 @@
-
-(the parentheses are necessary)
-
-## Defining Functors
-
-A functor with one argument can be defined like this:
-
-<!-- $MDX skip -->
-```ocaml
-module F (X : X_type) = struct
-  ...
-end
-```
-
-where `X` is the module that will be passed as argument, and `X_type` is its
-signature, which is mandatory.
-
-The signature of the returned module itself can be constrained, using this
-syntax:
-
-<!-- $MDX skip -->
-```ocaml
-module F (X : X_type) : Y_type =
-struct
-  ...
-end
-```
-
-or by specifying this in the `.mli` file:
-
-<!-- $MDX skip -->
-```ocaml
-module F (X : X_type) : Y_type
-```
-
-Overall, the syntax of functors is hard to grasp. The best may be to look at
-the source files
-[`set.ml`](https://github.com/ocaml/ocaml/blob/trunk/stdlib/set.ml) or
-[`map.ml`](https://github.com/ocaml/ocaml/blob/trunk/stdlib/map.ml) of the
-standard library.
+## Introduction
+
+In this tutorial, we look at how to apply functors and how to write functors. We also show some use cases involving functors.
+
+As suggested by the name, a _functor_ is almost like a function. However, while functions are between values, functors are between modules. A functor takes a module as a parameter and returns a module as a result. A functor in OCaml is a parametrised module, not to be confused with a [functor in mathematics](https://en.wikipedia.org/wiki/Functor).
+
+**Prerequisites**: [Modules](/docs/modules).
+
+## Project Setup
+
+This tutorial uses the [Dune](https://dune.build) build tool. Make sure you have installed version 3.7 or later. We start by creating a fresh project. We need a folder named `funkt` with files `dune-project`, `dune`, and `funkt.ml`.
+
+```shell
+$ mkdir funkt; cd funkt
+```
+
+Place the following in the file **`dune-project`**:
+```lisp
+(lang dune 3.7)
+(package (name funkt))
+```
+
+The content of the file **`dune`** should be this:
+```lisp
+(executable
+  (name funkt)
+  (public_name funkt)
+  (libraries str))
+```
+
+Create an empty file `funkt.ml`.
+
+Check that this works using the `dune exec funkt` command. It shouldn't do anything (the empty file is valid OCaml syntax), but it shouldn't fail either. The stanza `libraries str` makes the `Str` module (which we will use later) available.
+
+## Using an Existing Functor: `Set.Make`
+
+The standard library contains a [`Set`](/api/Set.html) module which is designed to handle sets. This module enables you to perform operations such as union, intersection, and difference on sets. You may check the [Set](/docs/sets) tutorial to learn more about this module, but it is not required to follow the present tutorial.
+
+To create a set module for a given element type (which allows you to use the provided type and its associated [functions](/api/Set.S.html)), it's necessary to use the functor `Set.Make` provided by the `Set` module. For reference only, here is a shortened version of the interface of `Set`:
+```ocaml
+module type OrderedType = sig
+  type t
+  val compare : t -> t -> int
+end
+
```

Comment:
@sabine Why did you remove this?
```ocaml
module type S = sig
  (** This is the module's signature returned by applying `Make` *)
end
```

Without it, understanding the signature of `Make` is a bit harder. Maybe we could have copied some `S`'s contents to make it easier to understand?